### ISSUE TITLE
Bug: Switch case statements allow use of undeclared variables

### ISSUE THREAD

**Issue by kamranayub** [Microsoft/TypeScript#18720]
Discovered a bug in my code and was surprised TS didn't catch it for me. It doesn't seem like this is allowed (in Chrome 60, at least).

**TypeScript Version:**  2.2.4

**Code**

```ts
var baz: number = 2
switch (baz) {
  case 1:
    let foo = 'bar'
    break;
  case 2:
    document.write(foo)
    break;
}
```

**Playground:** [link](http://www.typescriptlang.org/play/index.html#src=var%20baz%3A%20number%20%3D%202%0D%0Aswitch%20(baz)%20%7B%0D%0A%20%20case%201%3A%0D%0A%20%20%20%20let%20foo%20%3D%20'bar'%0D%0A%20%20%20%20break%3B%0D%0A%20%20case%202%3A%0D%0A%20%20%20%20document.write(foo)%0D%0A%20%20%20%20break%3B%0D%0A%7D)

**Expected behavior:**

TS should warn/error in second case statement that `foo` is not declared.

**Actual behavior:**

TS does not throw an error and `foo` is undefined

If this is expected/by design, I'm interested to know why (and if there's a way to catch it).

**Comment by DanielRosenwasser** [comment]
If you turn on `strictNullChecks` then TypeScript will correctly report that `foo` is used before it's assigned to.

**Comment by kamranayub** [comment]
So there are situations where this could be valid, I take it?

**Comment by kitsonk** [comment]
It is valid, since switch statements are a single block.  If you had another `let foo` in another case you would get a redeclaration error.  So string null checks is appropriate.

**Comment by gcnew** [comment]
My advice would be to use a block for every `case` statement that introduces new variables. The block will create a separate scope.

```ts
var baz: number = 2
switch (baz) {
    case 1: {
        let foo = 'bar'
        break;
    }
    case 2:
        document.write(foo)  // error as expected
        break;
}
```

**Comment by kamranayub** [comment]
Ah, okay, that makes more sense now! Thanks.

**Comment by mhegazy** [comment]
Automatically closing this issue for housekeeping purposes. The issue labels indicate that it is unactionable at the moment or has already been addressed.
